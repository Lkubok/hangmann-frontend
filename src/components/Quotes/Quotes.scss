@import "../../scss/mixins.scss";
@import "../../scss/variables.scss";

.contents-loading {
  @include flex-cc-row;
  height: 300px;
  p {
    flex-basis: 100%;
    text-align: center;
  }
}
.quotes-navigation {
  @include flex-sb-row;
  border-bottom: 1px solid $primary-color;
  margin-bottom: 0px;
  padding-bottom: 20px;
  padding-top: 10px;
}
.table-headings {
  font-weight: 700;
  margin-bottom: 25px;
}
.quote-table {
  display: flex;
  width: 100%;
  flex-flow: row wrap;

  &-body,
  &-header {
    flex-basis: 100%;
    tr {
      display: flex;
      flex-flow: row;
      justify-content: space-between;
      align-items: center;
      border-bottom: 1px solid $primary-color;
      padding-right: 10px;
      transition: all 0.4s ease-in-out;

      td:nth-child(2) {
        flex-basis: 50%;
        flex-grow: 0;
      }
      td:nth-child(3) {
        flex-basis: 20%;
      }
      td:nth-child(4),
      td:nth-child(5),
      td:nth-child(6),
      td:nth-child(7),
      td:nth-child(8) {
        flex-grow: 0;
        flex-basis: 7%;
      }
      td:nth-child(1) {
        flex-grow: 0;
        flex-basis: 3%;
      }
    }
    tr:nth-child(even) {
      background-color: $dark-color-light;
      &:hover {
        background-color: $dark-color-light-hover;
      }
    }

    tr:last-child {
      border: 0;
    }
  }
  &-header {
    border-bottom: 1px solid $primary-color;
    td {
      margin: 0;
      padding-top: 10px;
      padding-bottom: 20px;
    }
    tr {
      background-color: $primary-color;
      color: $dark-color;
    }
  }
  &-body {
    tr {
      &:hover {
        background-color: $dark-color-hover;
      }
    }
  }
}
