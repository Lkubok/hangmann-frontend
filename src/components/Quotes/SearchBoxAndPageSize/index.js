import SearchBoxAndPageSize from "./SearchBoxAndPageSize";
export default SearchBoxAndPageSize;
