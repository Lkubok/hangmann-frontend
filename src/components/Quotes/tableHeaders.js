export const tableHeaders = {
  number: "index",
  quote: "Quote",
  quoteAuthor: "Author",
  insertAuthor: "Inserted By",
  dateInsert: "Insert Date",
  dateModify: "Modify Date",
  lang: "Language",
  difficulty: "difficulty",
  modify: "Modify"
};
export const columnNames = Object.values(tableHeaders);
export const columnKeys = Object.keys(tableHeaders);
