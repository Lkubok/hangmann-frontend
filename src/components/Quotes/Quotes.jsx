import React, { Component } from "react";
import Pagination from "./Pagination";
import ButtonBox from "./ButtonBox";
import SearchBoxAndPageSize from "./SearchBoxAndPageSize";
import { connect } from "react-redux";
import Loading from "../EditQuote/Loading";
import {
  tableHeaders as columns,
  columnNames,
  columnKeys
} from "./tableHeaders";
import {
  updateQuotes,
  changeSorting,
  removeAllQuotes
} from "../../actions/quoteActions";
import * as selectors from "../../reducers/selectors";
import "./Quotes.scss";
const { REACT_APP_API_HOST } = process.env;

export class Quotes extends Component {
  constructor(props) {
    super(props);
    this.state = {
      columns
    };
  }
  componentDidMount() {
    if (this.props.quotes.length === 0)
      this.props.updateQuotes(REACT_APP_API_HOST);
  }
  componentWillUnmount() {
    this.props.removeAllQuotes();
  }

  handleClick = newSortByElem => () => {
    const { sortBy, sortOrder } = this.props;
    this.props.changeSorting(newSortByElem, sortBy, sortOrder);
  };

  renderTableHeads() {
    return columnNames.map(this.renderSingleHeader);
  }
  renderSingleHeader = (el, index) => {
    const { sortBy, sortOrder } = this.props;
    const key = columnKeys[index];
    const active = sortBy === key;
    return (
      <td className="table-headings" key={el} onClick={this.handleClick(key)}>
        {el}
        {active && (sortOrder === "ASC" ? "↑" : "↓")}
      </td>
    );
  };
  dateToHuman = dateStamp => {
    const date = new Date(dateStamp);
    const year = date.getFullYear().toString();
    return `${year.slice(-2)}-${date.getMonth() + 1}-${date.getDate()}`;
  };
  renderTableContents() {
    const { quotes, sortedQuotes, actualPage, pageLimit } = this.props;
    if (quotes.length === 0 || sortedQuotes.length === 0)
      return (
        <tr className="contents-loading">
          <td />
          <td>{quotes.length === 0 ? "Loading..." : "No match found"}</td>
        </tr>
      );
    const arrayStart = actualPage * pageLimit;
    const arrayToRender = sortedQuotes.slice(
      arrayStart,
      arrayStart + pageLimit
    );
    return arrayToRender.map((el, index) => (
      <tr className="quote-table-item" key={index}>
        <td>{index + arrayStart + 1}</td>
        <td>{el.quote}</td>
        <td>{el.quoteAuthor}</td>
        <td>{el.insertAuthor}</td>
        <td>{this.dateToHuman(el.dateInsert)}</td>
        <td>{this.dateToHuman(el.dateModify)}</td>
        <td>{el.lang}</td>
        <td>{el.difficulty}</td>
        <td>
          <ButtonBox
            canDelete={el.insertAuthor === this.props.userName ? true : false}
            id={el._id}
          />
        </td>
      </tr>
    ));
  }
  render() {
    const { quotes } = this.props;
    return (
      <>
        {quotes.length > 0 ? (
          <div className="quotes-wrapper">
            <div className="quotes-navigation">
              <Pagination />
              <SearchBoxAndPageSize />
            </div>
            <div className="quotes-body">
              <table className="quote-table">
                <thead className="quote-table-header">
                  <tr>{this.renderTableHeads()}</tr>
                </thead>
                <tbody className="quote-table-body">
                  {this.renderTableContents()}
                </tbody>
              </table>
            </div>
          </div>
        ) : (
          <Loading />
          // <div>a</div>
        )}
      </>
    );
  }
}

const mapStateToPops = (state, ownProps) => ({
  quotes: selectors.getQuotes(state),
  sortedQuotes: selectors.getSortedQuotes(state),
  actualPage: selectors.getActualPage(state),
  pageLimit: selectors.getPageLimit(state),
  sortOrder: selectors.getSortingOrder(state),
  sortBy: selectors.getSortingBy(state),
  userName: state.appParamsReducer.userName
});

const mapDispatchToProps = {
  updateQuotes,
  removeAllQuotes,
  changeSorting
};

export default connect(
  mapStateToPops,
  mapDispatchToProps
)(Quotes);
