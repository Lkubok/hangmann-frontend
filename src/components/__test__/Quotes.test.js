import React from "react";
import Enzyme, { shallow } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import { Quotes } from "../Quotes/Quotes";
Enzyme.configure({ adapter: new Adapter() });

describe("Quotes component", () => {
  let props;
  beforeEach(() => {
    props = {
      quotes: ["a", "b"],
      sortedQuotes: ["b", "c"],
      actualPage: 1,
      pageLimit: 5
    };
  });
  it("Should match snapshot", () => {
    const wrapper = shallow(<Quotes {...props} />);
    expect(wrapper).toMatchSnapshot();
  });
});
