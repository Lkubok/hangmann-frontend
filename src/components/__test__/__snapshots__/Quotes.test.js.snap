// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Quotes component Should match snapshot 1`] = `
<Fragment>
  <div
    className="quotes-wrapper"
  >
    <div
      className="quotes-navigation"
    >
      <Connect(Pagination) />
      <Connect(SearchBoxAndPageSize) />
    </div>
    <div
      className="quotes-body"
    >
      <table
        className="quote-table"
      >
        <thead
          className="quote-table-header"
        >
          <tr>
            <td
              className="table-headings"
              key="nr."
              onClick={[Function]}
            >
              nr.
            </td>
            <td
              className="table-headings"
              key="Quote"
              onClick={[Function]}
            >
              Quote
            </td>
            <td
              className="table-headings"
              key="Author"
              onClick={[Function]}
            >
              Author
            </td>
            <td
              className="table-headings"
              key="Ins By"
              onClick={[Function]}
            >
              Ins By
            </td>
            <td
              className="table-headings"
              key="Insert"
              onClick={[Function]}
            >
              Insert
            </td>
            <td
              className="table-headings"
              key="Modify"
              onClick={[Function]}
            >
              Modify
            </td>
            <td
              className="table-headings"
              key="Lang"
              onClick={[Function]}
            >
              Lang
            </td>
            <td
              className="table-headings"
              key="diff."
              onClick={[Function]}
            >
              diff.
            </td>
            <td
              className="table-headings"
              key="mod. "
              onClick={[Function]}
            >
              mod. 
            </td>
          </tr>
        </thead>
        <tbody
          className="quote-table-body"
        />
      </table>
    </div>
  </div>
</Fragment>
`;
